nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,741 @@
+        # which case we perform (maybe advanced) indexing on every element of
+        # the dictionnary.
+        idx = key
+        new_dict = type(self)(None)
+        try:
+            for k, v in self.items():
+                new_dict[k] = v[idx]
+        except (TypeError, ValueError, IndexError):
+            pass
+        else:
+            return new_dict
+
+        # Key was not a valid index/slice after all.
+        return self.store[key]  # Will raise the proper error.
+
+    def __delitem__(self, key):
+        del self.store[key]
+
+    def __iter__(self):
+        return iter(self.store)
+
+    def __len__(self):
+        return len(self.store)
+
+
+class PerArrayDict(SliceableDataDict):
+    """ Dictionary for which key access can do slicing on the values.
+
+    This container behaves like a standard dictionary but extends key access to
+    allow keys for key access to be indices slicing into the contained ndarray
+    values.  The elements must also be ndarrays.
+
+    In addition, it makes sure the amount of data contained in those ndarrays
+    matches the number of streamlines given at the instantiation of this
+    dictionary.
+    """
+    def __init__(self, nb_elements, *args, **kwargs):
+        self.nb_elements = nb_elements
+        super(PerArrayDict, self).__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        value = np.asarray(list(value))
+
+        if value.ndim == 1 and value.dtype != object:
+            # Reshape without copy
+            value.shape = ((len(value), 1))
+
+        if value.ndim != 2:
+            raise ValueError("data_per_streamline must be a 2D array.")
+
+        # We make sure there is the right amount of values
+        if self.nb_elements is not None and len(value) != self.nb_elements:
+            msg = ("The number of values ({0}) should match n_elements "
+                   "({1}).").format(len(value), self.nb_elements)
+            raise ValueError(msg)
+
+        self.store[key] = value
+
+
+class PerArraySequenceDict(SliceableDataDict):
+    """ Dictionary for which key access can do slicing on the values.
+
+    This container behaves like a standard dictionary but extends key access to
+    allow keys for key access to be indices slicing into the contained ndarray
+    values.  The elements must also be :class:`ArraySequence`.
+
+    In addition, it makes sure the amount of data contained in those array
+    sequences matches the number of elements given at the instantiation
+    of this dictionary.
+    """
+    def __init__(self, nb_elements, *args, **kwargs):
+        self.nb_elements = nb_elements
+        super(PerArraySequenceDict, self).__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        value = ArraySequence(value)
+
+        # We make sure there is the right amount of data.
+        if (self.nb_elements is not None and
+                value.nb_elements != self.nb_elements):
```

Comment:
If we think about what `ArraySequence` means it is an array of sequences. So calling `len(arr_seq)` gives you the number of sequences and `arr_seq.nb_elements` is the total number of sequences' items. What about: `sequences_(items|elements)_count` or `nb_sequences_(items|elements)`?